### ISSUE TITLE
Kafka messages aren't getting published

### ISSUE THREAD

**Issue by shivamka1** [gizmodemente/hello-kube#1]
I was trying this solution of yours wherein I was able to connect and make entry in the cassandra db and also was able to see kafka topic being created but for some reason I don't see any messages written to the kafka topic.

```
$ kafka-topics.sh --zookeeper localhost:2181 --list
__consumer_offsets
greetings
$ kafka-console-consumer.sh --topic greetings --bootstrap-server localhost:9092 --from-beginning
```

Could you please suggest what could be the reason?

P.S. This writing of messages to kafka topic was working in developement env `sbt runAll` but not in the kubernetes env.

**Comment by gizmodemente** [comment]
Hi friend,

The solution in kubernetes use headless services for Cassandra and Kafka
endpoints. To define a headless service you need to provide at least one ip
address in the endpoint object to allow kubernetes to find your Cassandra
or Kafka cluster (in this project this configuration could be found in
KafkaService.yaml file). Reading the Kafka cli commands in your mail, it
seems that you are using "localhost" as your Kafka address. That don't
going to work inside kubernetes using headless services.

I hope that this explanation can help you.

Thanks and best regards.

**Comment by shivamka1** [comment]
Thanks for your response.

The command below returns me the ip address of the host machine inside the minikube cluster which is what I have provided as the endpoints for both cassandra and kafka. This ip address seems to clearly work for cassandra but not for kafka.
```
minikube ssh "route -n | grep ^0.0.0.0 | awk '{ print \$2 }'"
```

I suspect that I need to change `localhost` to some ip address in the `server.properties` file but I am not able to figure out what IP address should I be placing there.
```
listeners=PLAINTEXT://localhost:9092
```

Please suggest.

**Comment by gizmodemente** [comment]
Yes, you need to configure the listeners in Kafka properties file with the
interfaces to listen. Maybe, for that configuration with minikube you can
try with 0.0.0.0 to listen in all interfaces. Don't use this configuration
in production.

I hope that this can help you.

**Comment by shivamka1** [comment]
When I tried `0.0.0.0` in the kafka properties file, I get the following error:
```
java.lang.IllegalArgumentException: requirement failed: 
advertised.listeners cannot use the nonroutable meta-address 0.0.0.0. Use a routable IP address.
```

May I know what IP address did you try in your setup for this yaml definition to work?

**Comment by gizmodemente** [comment]
I think that you need to understand what difference exists between
"listeners" property and "advertised.listeners" property in Kafka
configuration.

Maybe this article can help you to understand it and configure your
environment properly:

https://rmoff.net/2018/08/02/kafka-listeners-explained/

**Comment by shivamka1** [comment]
Thanks for sharing the link. It was very enlightening.

Since producer kafka client is not on the very same network as the brokers unlike consumer kafka client, we need to configure an additional listener like so:

```
listeners=INTERNAL://0.0.0.0:9093,EXTERNAL://0.0.0.0:9092
listener.security.protocol.map=INTERNAL:PLAINTEXT,EXTERNAL:PLAINTEXT
advertised.listeners=INTERNAL://localhost:9093,EXTERNAL://10.0.2.2:9092
inter.broker.listener.name=INTERNAL
```

We can verify messages in topic like so:
```
kafka-console-consumer.sh --bootstrap-server INTERNAL://0.0.0.0:9093 --topic greetings --from-beginning
{"name":"Alice","message":"Namastey"}
```

**Comment by gizmodemente** [comment]
Thank you for sharing your solution.
